### DISCUSSION TITLE
What's the point of the import step in resource migration?

### DISCUSSION THREAD

**Discussion by ppo-38** [Snowflake-Labs/terraform-provider-snowflake#3003] [Q&A] [answered]
Hello, 

I have a question about the necessity of the import step in the [resource migration guide ](https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/docs/technical-documentation/resource_migration.md):

**Here are the big steps in the guide (with no downtime)**
- List and remove legacy resources from the state
- Replace legacy resources by new resources in the terraform code
- Import the new resources into the state

**Goals**
In my configuration, I only need to replace `snowflake_role_grants` resources with `snowflake_grant_account_role` resources.

**Question**

In the final migration step, rather than importing the new resources into the state, couldn't we simply run an apply terraform? 

Since these are `GRANT ROLE` operations only, and there are no constraints on executing this type of command several times, performing an apply without first importing resources would also create the new resources in the state.

I tried this on a test environment and it worked well. The apply command creates the new resources in the terraform state and RBAC model is not modified on Snowflake. Next terraform plan is also ok : `Apply complete! Resources: 0 added, 0 changed, 0 destroyed.` 

So it seems to be transparent for users and I don't understand the point of the import step. Am I missing an important point?

Thanks for your help

**Comment by sfc-gh-asawicki** [comment, marked as answer]
Hey @ppo-38.

In your case, it should work like you described. It won't work for the objects where the underlying statements are not idempotent. So let's say you are migrating a warehouse, then the underlying command is CREATE, which will fail because the object already exists. That's where import comes in handy. We recommend it as a general route because it works in almost all situations without thinking about the underlying statements.

**Comment by ppo-38** [comment]
HI @sfc-gh-asawicki and thanks for your response ! 
I was rather convinced that apply would lead to the same result, but your answer reassures me for the next step.
